mobile/library/common/jni/types/Env.cc

```diff
@@ -0,0 +1,32 @@
+#include "library/common/jni/types/Env.h"
+
+#include "source/common/common/assert.h"
+
+#include "library/common/jni/jni_support.h"
+#include "library/common/jni/types/JavaVirtualMachine.h"
+
+namespace Envoy {
+namespace JNI {
+
+JNIEnv* Env::get() {
+  if (local_env_) {
+    return local_env_;
+  }
+
+  jint result = JavaVirtualMachine::getJavaVM()->GetEnv(reinterpret_cast<void**>(&local_env_),
+                                                        JavaVirtualMachine::getJNIVersion());
+  if (result == JNI_EDETACHED) {
+    // Note: the only thread that should need to be attached is Envoy's engine std::thread.
+    static const char* thread_name = "EnvoyMain";
+    JavaVMAttachArgs args = {JavaVirtualMachine::getJNIVersion(), const_cast<char*>(thread_name),
+                             nullptr};
+    result = attach_jvm(JavaVirtualMachine::getJavaVM(), &local_env_, &args);
+  }
+  RELEASE_ASSERT(result == JNI_OK, "Unable to get a JVM env for the current thread");
+  return local_env_;
+}
+
+thread_local JNIEnv* Env::local_env_ = nullptr;
```

Comment:
For this to detach on exit you need some RAII wrapper that calls detach in its dtor, similar to this https://github.com/jni-rs/jni-rs/blob/master/src/wrapper/java_vm/vm.rs#L523

Reply:
that thread_local is for `JNIEnv` - we use `static` for storing `JavaVM`.

Reply:
nvm - let my try something

Reply:
OK, I did attempt to go down this road but quickly ran into issues with the initialization of global variables / thread-safe in here:

```
[mutex.cc : 1404] RAW: Potential Mutex deadlock: 
        @ 0x2d8f14dd0 absl::DebugOnlyDeadlockCheck()
        @ 0x2d8f14f58 absl::Mutex::ReaderLock()
        @ 0x2d634cd1c absl::ReaderMutexLock::ReaderMutexLock()
        @ 0x2d634cb78 absl::ReaderMutexLock::ReaderMutexLock()
        @ 0x2d86c60b0 Envoy::Logger::DelegatingLogSink::log()
        @ 0x2d86bd970 spdlog::logger::sink_it_()
        @ 0x2d3c02328 spdlog::logger::log_it_()
        @ 0x2d81ba154 spdlog::logger::log_<>()
        @ 0x2d81b9eac spdlog::logger::log<>()
        @ 0x2d81b7c48 Envoy::BackwardsTrace::logFault()
        @ 0x2d81b7698 Envoy::SignalAction::sigHandler()
        @ 0x1873602a4 _sigtramp
        @ 0x18707da00 abort_with_payload_wrapper_internal
        @ 0x18707da34 dyld
        @ 0x18700c0a4 dyld4::CacheFinder::~CacheFinder()
        @ 0x18704027c dyld4::APIs::_tlv_atexit()
        @ 0x2d70115e4 __cxx_global_var_init
        @ 0x2d7011560 __tls_init
        @ 0x2d700f078 _ZTWN5Envoy3JNI3Env10local_env_E
        @ 0x2d700ec24 Envoy::JNI::Env::get()
```

 I say we move forward with what we have in here now and revisit this later on - I am at the point where I do not want to spend more time fighting c++ in here. If anything, I would prefer to address `call_method` comment before I fight thread_safe initialization.